### DISCUSSION TITLE
Discharge curve never calculated despite sensors and history

### DISCUSSION THREAD

**Discussion by mudge** [springfall2008/batpred#1299] [Q&A] [answered]
I've been running predbat for a few weeks now in "Control charge & discharge" mode but it never seems to generate a discharge curve with the following in `predbat.log` whenever I `touch apps.yaml`:

```
2024-07-12 13:00:11.828901: Find discharge curve with sensors sensor.givtcp_SERIAL_soc_kwh and number.givtcp_SERIAL_battery_discharge_rate and predbat.status and sensor.givtcp_SERIAL_battery_power
2024-07-12 13:00:13.977325: Find discharge curve has 8.0 days of data, max days 8
2024-07-12 13:00:14.017848: Note: Can not find battery charge curve (no final curve), one of the required settings for predbat_status, soc_kw, battery_power and charge_rate do not have history, check apps.yaml
```

(Where `SERIAL` is my real serial number.)

However, all of those sensors do exist and have values (according to Home Assistant's Developer Tools) and my apps.yaml has the following:

```yaml
  soc_kw:
    - sensor.givtcp_{geserial}_soc_kwh
  discharge_rate:
    - number.givtcp_{geserial}_battery_discharge_rate
  battery_power:
    - sensor.givtcp_{geserial}_battery_power
```

Am I missing something else that predbat needs to find the discharge curve?

**Comment by gcoan** [comment, marked as answer]
The battery charge and discharge curves can only be found when the battery is being fully charged or fully discharged respectively as the curve applies during the last 10% or so of the charge/discharge activity.  

In the last 8 days has your battery ever been fully discharged (down to 4%)?  If not this is probably why the curve can't be found.

**Reply by mudge** [reply]
I’m using the recommended “best SOC keep” of 0.5 kWh for my tariff (Octopus Flux) which is just over 5% of my 9.6 kWh battery. Looking at the plan over the next two days, the SOC % never goes below 6% (presumably for that reason) so presumably it’ll never get low enough for the curve to be calculated?

**Reply by gcoan** [reply]
It could be, the best soc keep is an advisory minimum so predbat can go through it but it will try to retain it. You can set the best soc keep to zero and see if the battery does discharge fully. In the summer if you've got a reasonable battery size it's quite possible that you won't exhaust the battery unless you deliberately try to do so with forced discharging. 
I wouldn't worry too much about this message anyway as the purpose of the charge and discharge curves is just so predbat can more accurately model how your battery behaves at the soc extremes. By default it assumes your battery charges and discharges linearly to top and bottom, and for the 9.5 that's pretty much the case for the discharge curve.
